File: lib/src/main/java/com/llama4j/RoPE.java
package com.llama4j;

final class RoPE {
    public static Pair<float[], float[]> precomputeFreqsCis(int contextLength, int headSize, double theta) {
        assert headSize % 2 == 0;
        int halfHead = headSize / 2;
        float[] cr = new float[contextLength * halfHead];
        float[] ci = new float[contextLength * halfHead];
        int n = 0;
        for (int pos = 0; pos < contextLength; ++pos) {
            for (int i = 0; i < headSize; i += 2) {
                float freq = (float) (1.0 / Math.pow(theta, i / (double) headSize));
                float val = pos * freq;
                cr[n] = (float) Math.cos(val);
                ci[n] = (float) Math.sin(val);
                n++;
            }
        }
        assert contextLength * halfHead == n;
        return new Pair<>(cr, ci);
    }

    public static Pair<float[], float[]> precomputeFreqsCisFromFreqs(int contextLength, int headSize, double ropeTheta, float[] ropeFreqFactors) {
        // freq_factors are divisors on top of the standard RoPE base frequencies:
        // theta_i = pos * (1 / (ropeTheta^(2i/headSize))) / freqFactors[i]
        int halfHead = ropeFreqFactors.length;
        assert halfHead == headSize / 2;
        float[] cr = new float[contextLength * halfHead];
        float[] ci = new float[contextLength * halfHead];
        int n = 0;
        for (int pos = 0; pos < contextLength; ++pos) {
            for (int i = 0; i < halfHead; i++) {
                float baseFreq = (float) (1.0 / Math.pow(ropeTheta, (2.0 * i) / headSize));
                float val = pos * baseFreq / ropeFreqFactors[i];
                cr[n] = (float) Math.cos(val);
                ci[n] = (float) Math.sin(val);
                n++;
            }
        }
        assert contextLength * halfHead == n;
        return new Pair<>(cr, ci);
    }
}

File: lib/src/main/kotlin/io/github/lbenedetto/internal/model/RoPE.kt
package io.github.lbenedetto.internal.model

import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin

object RoPE {
  fun precomputeFreqsCis(contextLength: Int, headSize: Int, theta: Double): Pair<FloatArray, FloatArray> {
    assert(headSize % 2 == 0)
    val halfHead = headSize / 2
    val cr = FloatArray(contextLength * halfHead)
    val ci = FloatArray(contextLength * halfHead)
    var n = 0
    for (pos in 0..<contextLength) {
      var i = 0
      while (i < headSize) {
        val freq = (1.0 / theta.pow(i / headSize.toDouble())).toFloat()
        val `val` = pos * freq
        cr[n] = cos(`val`.toDouble()).toFloat()
        ci[n] = sin(`val`.toDouble()).toFloat()
        n++
        i += 2
      }
    }
    assert(contextLength * halfHead == n)
    return Pair(cr, ci)
  }

  fun precomputeFreqsCisFromFreqs(
    contextLength: Int,
    headSize: Int,
    ropeTheta: Double,
    ropeFreqFactors: FloatArray
  ): Pair<FloatArray, FloatArray> {
    // freq_factors are divisors on top of the standard RoPE base frequencies:
    // theta_i = pos * (1 / (ropeTheta^(2i/headSize))) / freqFactors[i]
    val halfHead = ropeFreqFactors.size
    assert(halfHead == headSize / 2)
    val cr = FloatArray(contextLength * halfHead)
    val ci = FloatArray(contextLength * halfHead)
    var n = 0
    for (pos in 0..<contextLength) {
      for (i in 0..<halfHead) {
        val baseFreq = (1.0 / ropeTheta.pow((2.0 * i) / headSize)).toFloat()
        val `val` = pos * baseFreq / ropeFreqFactors[i]
        cr[n] = cos(`val`.toDouble()).toFloat()
        ci[n] = sin(`val`.toDouble()).toFloat()
        n++
      }
    }
    assert(contextLength * halfHead == n)
    return Pair(cr, ci)
  }
}
